### PR TITLE
ug: Add documentation for aktualizr-lite command line interface

### PR DESCRIPTION
**Pull request by detsch**
CLI documentation update for v95.

---
# PR Template and Checklist

Please complete as much as possible to speed up the reviewing process.

Readiness and adding reviewers as appropriate is required.

All PRs should be reviewed by a technical writer/documentation team and a peer.
If effecting customers—which is a majority of content changes—a member of Customer Success must also review.

## Readiness

* [ ] Merge (pending reviews)
* [ ] Merge after _date or event_
* [ ] Draft

## Overview

_Why merge this PR? What does it solve?_

## Checklist

* [x] Run spelling and grammar check, preferably with linter.
* [x] Avoid changing any header associated with a link/reference.
* [ ] Step through instructions (or ask someone to do so).
* [ ] Review for [wordiness](https://languagetool.org/insights/post/wordiness/)
* [ ] Match tone and style of page/section.
* [ ] Run `make linkcheck`.
* [ ] View HTML in a browser to check rendering.
* [ ] Use [semantic newlines](https://bobheadxi.dev/semantic-line-breaks/).
* [ ] follow best practices for commits.
  * [ ] Descriptive title written in the imperative.
  * [ ] Include brief overview of QA steps taken.
  * [ ] Mention any related issues numbers.
  * [ ] End message with sign off/DCO line (`-s, --signoff`).
  * [ ] Sign commit with your gpg key (`-S, --gpg-sign`).
  * [ ] Squash commits if needed.
* [ ] Request PR review by a technical writer and at least one peer.

## Comments

_Any thing else that a maintainer/reviewer should know._
_This could include potential issues, rational for approach, etc._

### CODE REVIEW

File: source/user-guide/custom-sota-client.rst [hunk 1]
```diff
@@ -101,3 +102,198 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The individual command line interface operations, specially `check`, `pull`, `install` and `run`, can be used to
+automate an update flow like to the one implemented by the main *aktualizr-lite* daemon, while allowing for limited
+customizations.
+
+.. highlight:: bash
+   :linenothreshold: 1
+
+Sample bash script illustrating usage of CLI operations and return codes handling::
```
Comment (mike-sul) [inline source/user-guide/custom-sota-client.rst:239]: I would move this script to the `sotactl` repo and reference it from this doc.

File: source/user-guide/custom-sota-client.rst [hunk 2]
```diff
@@ -101,3 +102,198 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+    - Failed to find the ostree commit and/or all Apps of the Target to be installed in the provided source bundle (offline mode only)
+- *11*: Failure
+    - Invalid TUF metadata
+- *12*: Failure
+    - TUF metadata is expired
+- *13*: Failure
+    - Unable to fetch TUF metadata
+- *14*: Failure
+    - TUF metadata not found in the provided path (offline mode only)
+- *15*: Failure
```
Comment (mike-sul) [inline source/user-guide/custom-sota-client.rst:167]: I think we need to ask someone from CS team to check if the CLI commands and return codes as explained here works for them.

File: source/user-guide/custom-sota-client.rst [hunk 3]
```diff
@@ -101,3 +102,198 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+Available commands:
+
+* ``list`` - lists the targets present in the currently stored TUF metadata.
+* ``check`` - updates the device's TUF repo with the latest Factory's TUF metadata, and lists the available targets.
+* ``pull`` - pulls the delta between the currently installed Target and the one specified with the `--update-name` option. If no target is specified, the latest one is used.
+* ``install`` - installs the previously pulled Target; yields an error if the specified Target has not been pulled before.
+* ``run`` - finalizes the installation of the Target; confirms an update after reboot on a new rootfs version and/or starts the updated apps. A ``finalize`` alias is provided for this command for backwards compatibility.
+* ``update`` - performs a complete update process, including the ``check``, ``pull`` and ``install`` operations. The ``run`` operation still needs to be manually invoked after the reboot.
+
+The commands set exit codes (``echo $?``) that can be used by the caller to act accordingly.
```
Comment (vanmaegima) [inline source/user-guide/custom-sota-client.rst:144]: This can be a new heading, similar to what we have with offline updates.

**Comment by vanmaegima** [comment]
I'm missing a note about stopping aklite daemon before running CLI.
We could cover:
- manual updates (stop aklite and stop fioconfig)
- skip daemon start from lmp-device-register
- skip daemon from the build (we have an example for custom-sota, what would be the alternative for CLI?)

**Comment by vanmaegima** [comment]
I haven't reviewed return codes and sample script yet.

File: source/user-guide/custom-sota-client.rst (hunk 1 above)
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:239]: I was thinking about having it on a repo an also keep it in the docs, I believe it is simply enough to contribute to the docs context.

File: source/user-guide/custom-sota-client.rst (hunk 3 above)
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:144]: Done.

File: source/user-guide/custom-sota-client.rst [hunk 4]
```diff
@@ -101,3 +102,198 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
 * ``pull`` - pulls the delta between the currently installed and the specified one.
 * ``install`` - installs the previously pulled Target; yields an error if the specified Target has not been pulled before.
 * ``run`` - finalizes the installed Target; confirms an update after reboot on a new rootfs version and/or starts the updated apps.
+
+Command Line Interface (experimental)
+-------------------------------------
+The `aktualizr-lite` daemon executable can be invoked to perform individual operations allowing more control over the update flow.
+This interface is subject to change over the next releases.
+
+.. prompt:: bash
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:111]: Changed.

File: source/user-guide/custom-sota-client.rst [hunk 5]
```diff
@@ -101,3 +102,198 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
 * ``pull`` - pulls the delta between the currently installed and the specified one.
 * ``install`` - installs the previously pulled Target; yields an error if the specified Target has not been pulled before.
 * ``run`` - finalizes the installed Target; confirms an update after reboot on a new rootfs version and/or starts the updated apps.
+
+Command Line Interface (experimental)
+-------------------------------------
+The `aktualizr-lite` daemon executable can be invoked to perform individual operations allowing more control over the update flow.
+This interface is subject to change over the next releases.
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:109]: Updated format to bold

**Comment by doanac** [comment]
Docs for 4564d81bbd36101ef653fd0fe339153bff190d83 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2717/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for 5e568db8bb424cebb1d1ce6084a4bf23e83025d1 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2742/docs/artifacts/html/index.html

File: source/user-guide/custom-sota-client.rst [hunk 6]
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
 * ``run`` - finalizes the installed Target; confirms an update after reboot on a new rootfs version and/or starts the updated apps.
+
+Command Line Interface - CLI (Aktualizr-lite Manual Mode)
+---------------------------------------------------------
+
+The `aktualizr-lite` executable can be invoked to perform individual operations allowing more control over the update flow.
+
+.. warning:: The Command Line Interface is on beta stage, and is subject to change over the next releases.
+
+.. note:: In order to use the run individual `aktualizr-lite` commands, the `aktualizr-lite` service needs to be stopped with ``sudo systemctl stop aktualizr-lite`` and/or disabled with ``sudo systemctl disable aktualizr-lite``
```
Comment (vanmaegima) [inline source/user-guide/custom-sota-client.rst:113]: I wonder if we need to also stop fioconfig in this case as iiuc it also starts aklite daemon

Comment (detsch) [reply source/user-guide/custom-sota-client.rst:113]: Yes, but perhaps we should adjust this behavior, otherwise it is a pain to use fioconfig with any custom OTA solution. I'm checkin what can we do about it, probably adjusting the `fioconfig` toml updates handler.

**Comment by doanac** [comment]
Docs for 15489180d213f5cd85017a9b9fc156605f26699a are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2743/docs/artifacts/html/index.html

File: source/user-guide/custom-sota-client.rst (hunk 6 above)
Comment (vanmaegima) [reply source/user-guide/custom-sota-client.rst]: awesome, thanks!

Comment (detsch) [reply source/user-guide/custom-sota-client.rst]: It is now merged, disabling aklite service is now enough. Fioconfig won't (re)start it anymore, even after changes in device config, such as apps list.

File: source/user-guide/custom-sota-client.rst [hunk 7]
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+
+   sudo aktualizr-lite list
+
+Both commands can be used in conjunction with the ``--json 1`` option, which will change the output format to JSON, and will by default omit other log outputs.
+
+
+Apply Update
+^^^^^^^^^^^^
+
+The ``update`` command applies the latest available update to the device, after updating the TUF metadata and pulling the required data.
```
Comment (mike-sul) [inline source/user-guide/custom-sota-client.rst:164]: This `pulling the required data` is a bit confusing. IMO, it can be removed.

Comment (mike-sul) [inline source/user-guide/custom-sota-client.rst:164]: Maybe "command pulls and installs" instead of "applies" which doesn't not necessary implies downloading.

File: source/user-guide/custom-sota-client.rst [hunk 8]
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+^^^^^^^^^^^^^^^^^^^
+
+To view the current status of the device::
+
+    sudo aktualizr-lite status
+
+Fetch TUF Metadata and List Updates
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``check`` command will refresh the Targets metadata from the OTA server, and present you with a list of available Targets::
```
Comment (mike-sul) [inline source/user-guide/custom-sota-client.rst:150]: @kprosise Do you think this "refresh the Targets metadata from the OTA server" means what the command actually does?

File: source/user-guide/custom-sota-client.rst [hunk 9]
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+The ``update`` command applies the latest available update to the device, after updating the TUF metadata and pulling the required data.
+This includes both OSTree and Docker app Targets::
+
+   sudo aktualizr-lite update
+
+To update to a specific build number or target name, the ``--update-name`` option can be used::
+
+   sudo aktualizr-lite update --update-name <build_number_or_name>
+
+A reboot command will be required after applying an update, followed by the execution on the  ``run`` command to finalize the update process::
```
Comment (mike-sul) [inline source/user-guide/custom-sota-client.rst:173]: I think the term "apply" is not used consistently across this doc. In one case it means an overall update process, in another like this it means installation.

File: source/user-guide/custom-sota-client.rst [hunk 10]
```diff
@@ -101,3 +102,255 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+
+   sudo aktualizr-lite run
+
+
+.. note::
+   An update can only be performed when the original and update Targets are under the same tag.
+   In case the update is tagged differently, it is required to switch tags before running this command.
```
Comment (mike-sul) [inline source/user-guide/custom-sota-client.rst:180]: `the update` --> `the Target`?

**Comment by doanac** [comment]
Docs for 9eebdc64277c4bc58fce8eac4700081894bfc394 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2751/docs/artifacts/html/index.html

**Comment by detsch** [comment]
> I'm missing a note about stopping aklite daemon before running CLI. We could cover:
> 
> * manual updates (stop aklite and stop fioconfig)
Added note.

> * skip daemon start from lmp-device-register
Added note.

> * skip daemon from the build (we have an example for custom-sota, what would be the alternative for CLI?)
We do need the daemon binary, so the build itself is not changed someone is going to use CLI. Perhaps adding something about not enabling the aklite service by default in the build? Not sure hot that can be done, will have to check it.

File: source/user-guide/custom-sota-client.rst (hunk 10 above)
Comment (detsch) [reply source/user-guide/custom-sota-client.rst]: Updated.

File: source/user-guide/custom-sota-client.rst (hunk 9 above)
Comment (detsch) [reply source/user-guide/custom-sota-client.rst]: Updated.

File: source/user-guide/custom-sota-client.rst (hunk 7 above)
Comment (detsch) [reply source/user-guide/custom-sota-client.rst]: Updated.

Comment (detsch) [reply source/user-guide/custom-sota-client.rst]: Removed.

**Comment by doanac** [comment]
Docs for 968cbeef7cbe4976706db1a0f6869a173a489a44 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2752/docs/artifacts/html/index.html

File: source/user-guide/custom-sota-client.rst [hunk 11]
```diff
@@ -101,3 +102,217 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+A reboot command will be required after installing an update,
+followed by the execution on the  ``run`` command to finalize the update process::
+
+   sudo aktualizr-lite run
+
+
+.. note::
+   An update can only be performed when the original and update Targets are under the same tag.
+   In case the Target is tagged differently,
+   it is required to switch tags before running this command.
```
Comment (mike-sul) [inline source/user-guide/custom-sota-client.rst:195]: I would rather recommend to add the device's tag to the Target, switching tags on a device is rather the last resort.

Maybe we should remove this note at all since this is a general rule, not specific to the CLI, is applicable to any types of update.

**Review by mike-sul — APPROVED** [review]
LGTM, just minor comments.

File: source/user-guide/custom-sota-client.rst (hunk 11 above)
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:195]: I agree. Having it here only adds uneccessary complexity to the documentation. The target would not be listed in the `list` and `check` commands anyway. Removing the note.

**Comment by doanac** [comment]
Docs for f227e917db8d0b406d3e6c2e657a04a4501d9788 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2753/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for c9b613078f1d8efd6375abf6db3e1dab203f25de are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2758/docs/artifacts/html/index.html

**Comment by detsch** [comment]
@kprosise can you take a look? Feel free to do text adjustments and merge if no changes are required from our side. Thanks!

**Review by kprosise — COMMENTED** [review]
Made some minor suggestions, but otherwise it LGTM

**Comment by kprosise** [comment]
@detsch, if you could add the suggestions, if they seem appropriate, then squash the commits, I will then go ahead and merge. Thank you!

File: source/user-guide/custom-sota-client.rst [hunk 12]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+- *110*: Failure
+    - Online installation failed, rollback performed
+- *130*: Failure
+    - Installation failed and rollback operation was not successful
+
+Automating the use of CLI operations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The individual command line interface operations,
+specially ``check``, ``pull``, ``install`` and ``run``,
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:304]: Applied.

File: source/user-guide/custom-sota-client.rst [hunk 13]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+The ``install``  command installs the Target, which should have been previously pulled.
+It yields an error if the specified Target has not been pulled before, and also supports the ``--update-name`` option.
+
+It is necessary to verify the return codes for each command in order to guarantee the correct update process flow,
+as detailed in the next section.
+
+Exit Codes
+^^^^^^^^^^
+The commands set exit codes (``echo $?``) that can be used by the caller to act accordingly.
+The possible return codes for the CLI commands are listed bellow:
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:216]: Applied.

File: source/user-guide/custom-sota-client.rst [hunk 14]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+
+The ``install``  command installs the Target, which should have been previously pulled.
+It yields an error if the specified Target has not been pulled before, and also supports the ``--update-name`` option.
+
+It is necessary to verify the return codes for each command in order to guarantee the correct update process flow,
+as detailed in the next section.
+
+Exit Codes
+^^^^^^^^^^
+The commands set exit codes (``echo $?``) that can be used by the caller to act accordingly.
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:215]: Applied.

File: source/user-guide/custom-sota-client.rst [hunk 15]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+
+The ``check`` command updates the Targets metadata.
+
+The ``pull`` command pulls the delta between the currently installed Target and the one specified with the ``--update-name`` option.
+If no target is specified, the latest one is used.
+
+The ``install``  command installs the Target, which should have been previously pulled.
+It yields an error if the specified Target has not been pulled before, and also supports the ``--update-name`` option.
+
+It is necessary to verify the return codes for each command in order to guarantee the correct update process flow,
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:210]: Applied.

File: source/user-guide/custom-sota-client.rst [hunk 16]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+   sudo aktualizr-lite run
+
+
+.. warning::
+   Downgrading to a older Target is neither recommended or supported by our team;
+   doing so may lead to unverified corner cases.
+   Only choose to do so mindfully.
+   For any update, always test before rolling out to production devices.
+
+Aktualizr-lite command line interface also allows the update steps to be performed individually,
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:198]: Applied.

File: source/user-guide/custom-sota-client.rst [hunk 17]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+Apply Update
+^^^^^^^^^^^^
+
+The ``update`` command pulls and installs the latest available update to the device,
+after updating the TUF metadata.
+This includes both OSTree and Docker app Targets::
+
+   sudo aktualizr-lite update
+
+To update to a specific build number or target name,
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:181]: Applied.

File: source/user-guide/custom-sota-client.rst [hunk 18]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+
+
+View Current Status
+^^^^^^^^^^^^^^^^^^^
+
+To view the current status of the device::
+
+    sudo aktualizr-lite status
+
+Fetch TUF Metadata and List Updates
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:154]: Applied.

File: source/user-guide/custom-sota-client.rst [hunk 19]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+.. warning:: The Command Line Interface is on beta stage,
+    and is subject to change over the next releases.
+
+.. note:: In order to use the run individual `aktualizr-lite` commands,
+    the `aktualizr-lite` service needs to be stopped with ``sudo systemctl stop aktualizr-lite``
+    and/or disabled with ``sudo systemctl disable aktualizr-lite``.
+
+.. note:: If lmp-device-register is used,
+    the `--start-daemon 0` is recommended
+    in order to avoid starting aktualizr-lite daemon automatically.
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:120]: Applied.

File: source/user-guide/custom-sota-client.rst [hunk 20]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+
+.. warning:: The Command Line Interface is on beta stage,
+    and is subject to change over the next releases.
+
+.. note:: In order to use the run individual `aktualizr-lite` commands,
+    the `aktualizr-lite` service needs to be stopped with ``sudo systemctl stop aktualizr-lite``
+    and/or disabled with ``sudo systemctl disable aktualizr-lite``.
+
+.. note:: If lmp-device-register is used,
+    the `--start-daemon 0` is recommended
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:119]: Applied.

File: source/user-guide/custom-sota-client.rst [hunk 21]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+Command Line Interface - CLI (Aktualizr-lite Manual Mode)
+---------------------------------------------------------
+
+The `aktualizr-lite` executable can be invoked to perform individual operations allowing more control over the update flow.
+
+.. warning:: The Command Line Interface is on beta stage,
+    and is subject to change over the next releases.
+
+.. note:: In order to use the run individual `aktualizr-lite` commands,
+    the `aktualizr-lite` service needs to be stopped with ``sudo systemctl stop aktualizr-lite``
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst:115]: Applied.

File: source/user-guide/custom-sota-client.rst [hunk 22]
```diff
@@ -101,3 +102,209 @@ In addition to the default daemon mode, users can run it as a CLI utility and pe
...
+- *40*: Failure
+    - No pending installation to run
+- *99*: Failure
+    - Offline installation failed, rollback performed
+- *110*: Failure
+    - Online installation failed, rollback performed
+- *130*: Failure
+    - Installation failed and rollback operation was not successful
+
+Automating the use of CLI operations
```
[the remark this answers is not in the document]
Comment (detsch) [reply source/user-guide/custom-sota-client.rst]: Applied.

**Comment by detsch** [comment]
Thanks @kprosise! Applied all your suggestions, and rebased to the latest `main` . Feel free to merge when appropriate.

**Comment by doanac** [comment]
Docs for 9d9037ddd92af8a25048bfab8a839080a23ad29d are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2764/docs/artifacts/html/index.html
